### PR TITLE
Fix/inlane double lane

### PR DESCRIPTION
**Pull request by MishkaMN**
# PR Details
## Description

This fixes logic in inlanecruising_plugin where it includes both of previous adjacent lanelets used for lanechanging in its curvature calculation logic. It only includes the lanelets in the straight part of the road (doesn't include adjacent lanelets that would have occured in the future as well).
Other changes include:
- ... TODO

This branch's changes in route_following_plugin should be ignored. However, until other PR in https://github.com/usdot-fhwa-stol/carma-platform/pull/1272 is merged, we are keeping the changes just for completeness.

Note:
Currently it is pointing to temporary branch named fix/zephyr-revert in order to better check difference as release/zephyr branch is corrupted.

## Motivation and Context

See above.

## How Has This Been Tested?
Integration tested on Blue Lexus

## Types of changes

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [X] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp [hunk 1]
```diff
@@ -572,22 +572,65 @@ std::vector<PointSpeedPair> InLaneCruisingPlugin::maneuvers_to_points(const std:
...
+    }
+
     ROS_DEBUG_STREAM("Maneuver");
 
     lanelet::BasicLineString2d downsampled_centerline;
     downsampled_centerline.reserve(200);
 
-    for (auto l : lanelets)
+    // getLaneletsBetween returns shortest path that is ordered
+    // exclude lanechanges and plan for only the straight part
```
Comment (JonSmet) [inline inlanecruising_plugin/src/inlanecruising_plugin.cpp:587]: Can we add a comment to explain how/why a lane change may exist within inlanecruising plugin's maneuvers_to_points() function?

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp [hunk 2]
```diff
@@ -616,8 +659,40 @@ std::vector<PointSpeedPair> InLaneCruisingPlugin::maneuvers_to_points(const std:
...
+        continue;
+      }
+
+      double delta_d = lanelet::geometry::distance2d(prev_point, current_point);
+      dist_accumulator += delta_d;
+
+      double ending_downtrack = maneuvers.back().lane_following_maneuver.end_dist;
+      if(ending_downtrack + config_.buffer_ending_downtrack < wm_->getRouteEndTrackPos().downtrack){
+        ending_downtrack = ending_downtrack + config_.buffer_ending_downtrack;
+      }
```
Comment (msmcconnell) [inline inlanecruising_plugin/src/inlanecruising_plugin.cpp:681]: This ending downtrack change should be before this for loop as it returns a constant value. No need to have it in the loop.

**Review by JonSmet — CHANGES_REQUESTED** [review]

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp [hunk 3]
```diff
@@ -616,8 +659,40 @@ std::vector<PointSpeedPair> InLaneCruisingPlugin::maneuvers_to_points(const std:
...
+      prev_point = current_point;
+      i++;
+    }
+    if (max_i == 0) {
+      max_i = points_and_target_speeds.size() - 1;
+    }
 
-  return points_and_target_speeds;
+    std::vector<PointSpeedPair> constrained_points(points_and_target_speeds.begin(), points_and_target_speeds.begin() + max_i);
```
Comment (msmcconnell) [inline inlanecruising_plugin/src/inlanecruising_plugin.cpp:694]: There needs to be a way to remove the extra points added just for supporting the spline fit after computing the curvature. This is what we discussed and you implemented in the car on 5/14 I believe.

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp [hunk 4]
```diff
@@ -572,22 +579,65 @@ std::vector<PointSpeedPair> InLaneCruisingPlugin::maneuvers_to_points(const std:
...
+      straight_lanelets.push_back(lanelets[curr_idx]);
+
+      // add all lanelets on the straight road until next lanechange
+      while (curr_idx + 1 < lanelets.size() && 
+              std::find(following_lanelets.begin(),following_lanelets.end(), lanelets[curr_idx + 1]) != following_lanelets.end())
+      {
+        curr_idx++;
+        ROS_DEBUG_STREAM("Added following lanelet Id: " << lanelets[curr_idx].id());
+        straight_lanelets.push_back(lanelets[curr_idx]);
+        following_lanelets = wm->getMapRoutingGraph()->following(lanelets[curr_idx]);
```
Comment (msmcconnell) [inline inlanecruising_plugin/src/inlanecruising_plugin.cpp:626]: Using the RoutingGraph does not mean you will only get lanelets on the shortest path. I think you would also need to check if the returned following lanelets were in the shortest path. No?

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp (hunk 1 above)
Comment (MishkaMN) [reply inlanecruising_plugin/src/inlanecruising_plugin.cpp:594]: Thanks! I added extra comments to clarify

File: inlanecruising_plugin/src/inlanecruising_plugin.cpp (hunk 3 above)
Comment (MishkaMN) [reply inlanecruising_plugin/src/inlanecruising_plugin.cpp:711]: we are dropping the extra points around the line which has comment `//drop buffer points here`

**Comment by MishkaMN** [comment]
closed this PR as new one is updated here https://github.com/usdot-fhwa-stol/carma-platform/pull/1307

**Pull request closed without merging**